project/conventions.md

```diff
@@ -0,0 +1,108 @@
+the SPIRE Helm Chart project, permitting multiple developers to align
+to a similar style and standard when offering their contributions.
+
+It is impossible to describe every scenario a developer might encounter,
+likewise it is equally impossible to describe the best choice to make
+under all future circumstances.  For this reason, many conventions also
+have a small explanation of the rationale behind the guidance, allowing
+future development efforts to weigh and balance the conflicting forces
+in ambiguous situations not clearly covered.
+
+This document is intended to be a living document.  If you have an issue
+with a convention, we hope that you will create an issue describing the
+problem, and a pull request suggesting the update.  By addressing the
+convention, we hope to avoid arguments about conventions in the pull
+request review process, which slow the process and are often subject to
+the popular whims of the moment.
+
+## Conventions used in this document
+
+While this is not a software specification, the key words "MUST",
+"MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT",
+"RECOMMENDED",  "MAY", and "OPTIONAL" in this document are to be
+interpreted as closely to the convention as it would be to a
+specification described in RFC 2119.
+
+Use of the word in its non-all caps form SHOULD NOT be interpreted in
+the strict definitions user RFC 2119, but SHOULD be interpreted in
+context with a meaning aligned to RFC 2119.
+
+## Helm Chart conventions
+
+Helm values are organized in a tree structure, with each entry having
+a key, the path within the structure, and a value, the contents held
+at that path.  Some of the contents include objects and arrays, which
+are structures that permit multiple values under a particular path.
+
+This design choice mimics the representation of a file system, and we
+may use file system analogies to describe the conventions and choices
+in the project.
+
+### Preference for absolute paths in templates
+
+Outside of looping constructs that require relative paths, templates
+should use absolute paths (paths starting with the `.` character) for
+their value references.  
+
+### Helm Key path elements
+
+Elements refer to items in the configuration.  Some of those items are
+external to the project, while others are components the project choose
+to implement the Helm Chart packaging solution.
+
+### Preferred characters for helm path elements
+
+Elements within a helm chart path SHOULD limit the use of special
+characters, even when such characters are supported by Helm.  ASCII
+letters, upper and lower case, ASCII numerals, and the ASCII underscore
+character are always permitted.  
+
+Other characters, such as `@`, `~`, `-`, `&`, `+` are often used as
+shorthand to represent, respectively `at`, `similar / not`, 
+`minus / without`, `and / reference`, `plus / with`.  Helm key elements
+MUST NOT contain such symbols, as it burdens the reader to deduce the
+intent of the writer.
```

Comment:
'-' is often used as ' ' as well, and one of the few items that is dns compatible for naming k8s objects. in some cases its one of the few options and we should be a little careful preventing it.

Reply:
They are more difficult to deal with in helm templates, yes. But they will be unavoidable in cases where nested charts may be used or in cases where we want to use the same key as the plugin name, for example, from the spire settings. So, 'MUST NOT' is probably too strong a wording with regard to '-'.

Reply:
@kfox1111 These aren't laws.  They're conventions.  If you can get everyone on board with an exception, then that's fine by me.  Trying to make them into laws will only mean we will never have a guideline document that says what we prefer / desire; because it will be littered with far too many rules to be followed.

Reply:
Just trying to understand. There was seemly great care paid attention to specify terms to match RFC 2119, then the strongest form used MUST NOT used rather then, say, "SHOULD NOT', which clearly articulates that there is room for reasonable exceptions.

Reply:
**If you think I used the wrong phrase, I'm open to a request to change it.**

I personally don't want symbols that have two meanings, as `message2Route` can always be represented as `messageTooRoute` or `messageTwoRoute` which makes the context of the previous `2` obvious.  The same goes for the other symbols above.  If you want to open the door to them sometimes being used, I'm ok with that, getting a document into the repo we can reference more important.

After all, we can always then do an Issue to clarify / change / rewrite items we disagree with, as we find the scenarios that make the current document insufficient.  But without a document to change, there's just nothing to refine.  That said, I'd want the document that goes in to be mostly ok with the team, it has no value if nobody agrees to it.

Reply:
@kfox1111 If I wasn't clear on how I would like the request to change it.  Use the review tools to provide the line as you want it written, so I don't have to guess at the desired edit a few times.

Reply:
@kfox1111 In rereading this I see the main point was the strength of the guideline, not the content of the guideline.

I'm of the mind that for every rule, there is a scenario where an exception can reasonably occur.  Because of this viewpoint, I don't try to make rules for others absolute.  If you want this to be absolute, I can change it (but I personally think people will do the right thing without it being made absolute).

Do you want it changed?  It's not a big deal to me.